In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.svm import LinearSVC
import warnings
import sklearn.svm

In [2]:
data_df = pd.read_excel('LSVT_voice_rehabilitation.xlsx')


In [3]:
X = data_df.to_numpy()
Y_data = pd.read_excel('LSVT_voice_rehabilitation.xlsx', sheet_name='Binary response')

# Linear kernel

In [4]:
warnings.filterwarnings('ignore')
clf = LinearSVC(max_iter=50000)
Y = Y_data.to_numpy()
Y = np.squeeze(Y)
f1_scores = cross_val_score(clf, X, Y, cv=10, scoring='f1_macro')
acc_scores = cross_val_score(clf, X, Y, cv=10)
print(f'F1-measure score for linear svm is: {np.mean(f1_scores)}')
print(f'Accuracy for linear svm is: {np.mean(acc_scores)}')

F1-measure score for linear svm is: 0.38947330447330447
Accuracy for linear svm is: 0.5185897435897436


# Polynomial kernel

In [5]:
for d in (2, 3, 5):
    for r in (-5, 0, 10):
        clf = sklearn.svm.SVC(kernel='poly', degree=d, coef0=r, max_iter=50000)
        f1_scores = cross_val_score(clf, X, Y, cv=10, scoring='f1_macro')
        acc_scores = cross_val_score(clf, X, Y, cv=10)
        print(f'F1-measure score for polynomial svm with d={d}, r={r} is: {np.mean(f1_scores)}')
        print(f'Accuracy for polynomial svm with d={d}, r={r} is: {np.mean(acc_scores)}')
        print()

F1-measure score for polynomial svm with d=2, r=-5 is: 0.39535383862783247
Accuracy for polynomial svm with d=2, r=-5 is: 0.4948717948717949

F1-measure score for polynomial svm with d=2, r=0 is: 0.3998268398268398
Accuracy for polynomial svm with d=2, r=0 is: 0.6666666666666667

F1-measure score for polynomial svm with d=2, r=10 is: 0.3998268398268398
Accuracy for polynomial svm with d=2, r=10 is: 0.6666666666666667

F1-measure score for polynomial svm with d=3, r=-5 is: 0.44493359366811386
Accuracy for polynomial svm with d=3, r=-5 is: 0.5128205128205128

F1-measure score for polynomial svm with d=3, r=0 is: 0.3998268398268398
Accuracy for polynomial svm with d=3, r=0 is: 0.6666666666666667

F1-measure score for polynomial svm with d=3, r=10 is: 0.3998268398268398
Accuracy for polynomial svm with d=3, r=10 is: 0.6666666666666667

F1-measure score for polynomial svm with d=5, r=-5 is: 0.49150814871403103
Accuracy for polynomial svm with d=5, r=-5 is: 0.532051282051282

F1-measure scor

# rbf kernel

In [11]:
for gamma in (0.001, 2.0, 300.0):
    clf = sklearn.svm.SVC(kernel='rbf', gamma=gamma)
    f1_scores = cross_val_score(clf, X, Y, cv=10, scoring='f1_macro')
    acc_scores = cross_val_score(clf, X, Y, cv=10)
#     print(acc_scores)
    print(f'F1-measure score for rbf svm with gamma={clf.gamma} is: {np.mean(f1_scores)}')
    print(f'Accuracy for rbf svm with gamma={clf.gamma} is: {np.mean(acc_scores)}')
    print()

F1-measure score for rbf svm with gamma=0.001 is: 0.3998268398268398
Accuracy for rbf svm with gamma=0.001 is: 0.6666666666666667

F1-measure score for rbf svm with gamma=2.0 is: 0.3998268398268398
Accuracy for rbf svm with gamma=2.0 is: 0.6666666666666667

F1-measure score for rbf svm with gamma=300.0 is: 0.3998268398268398
Accuracy for rbf svm with gamma=300.0 is: 0.6666666666666667



# Sigmoid kernel

In [14]:
for r in (0.001, 0.5, 2.0):
    clf = sklearn.svm.SVC(kernel='sigmoid', coef0=r)
    f1_scores = cross_val_score(clf, X, Y, cv=10, scoring='f1_macro')
    acc_scores = cross_val_score(clf, X, Y, cv=10)
#     print(acc_scores)
    print(f'F1-measure score for sigmoid svm with r={clf.coef0} is: {np.mean(f1_scores)}')
    print(f'Accuracy for sigmoid svm with r={clf.coef0} is: {np.mean(acc_scores)}')
    print()

F1-measure score for sigmoid svm with r=0.001 is: 0.49537931839402427
Accuracy for sigmoid svm with r=0.001 is: 0.5743589743589744

F1-measure score for sigmoid svm with r=0.5 is: 0.38422417451829216
Accuracy for sigmoid svm with r=0.5 is: 0.5269230769230769

F1-measure score for sigmoid svm with r=2.0 is: 0.3998268398268398
Accuracy for sigmoid svm with r=2.0 is: 0.6666666666666667



# تاثیر پارامتر های هر مدل به عملکرد آن:
## کرنل خطی: پارامتری نداشت.
## کرنل چند جمله ای:
    دو پارامتر d,r را داشت که به ازای هر پارامتر 3 اندازه متفاوت تست شد. تاثیر پارامتر r به این صورت بود که معمولا حالت r=0 خوب عمل نمیکرد و حالت -5 نیز خوب عمل میکرد.در حالت درجه5 نیز r برابر با 10 خوب عمل کرد.برای درجه های تست شده (2و3و5) با بالا رفتن درجه نیز عملکرد الگوریتم به طور نسبی بهتر در f1-measure و بدتر در accuracy شد.
## کرنل سیگموید:
    تنها پارامتر r را داشت که برای سه مقدار تست شده بهترین عملکرد  را r برابر با 0.001 داشت از نظر f1-measure و r=2 از نظر accuracy.
## کرنل rbf:
    پارامتر آن گاما است که با تغییر گاما و قرار دادن مقادیر مختلف برای آن عملکرد الگوریتم دچار تغییر نشد!

# بهترین مدل:
### بهترین مدل از نظر اندازه F1 مدل های با کرنل سیگموید و اندازه r برابر 0.001 و مدل چندجمله ای با d=5,r=-5 بود که مقداری به اندازه 0.49 می‌شد.
### بهترین مدل از نظر اندازه accuracy برابر با مدل سیگویید با r=2، مدل های با کرنل rbf و چهار مدل در کرنل چند جمله به ترتیب با اندازه های d,r برابر با 2,0 ، 2,10 ، 3,0 ، 3,10 است که مقدار 0.66 م‌شد.
